### ISSUE TITLE
Better Recommendations🧙‍♂️

### ISSUE THREAD

**Issue by Praful932** [Praful932/Kitabe#38]
**Is your feature request related to a problem? Please describe.**
There are cases when there is not enough books to recommend the user, In this cases currently next top books are taken up according to TF-IDF approach.

**Describe the solution you'd like**
A better approach would be to take in equally from TF-IDF also similar genre books. If the remaining books to be recommended are 4, 2 - TF-IDF & 2 - Similar genres book. Also take into account already rated & included books before adding those 2 similar genres book.

[Concerning Code Line](https://github.com/Praful932/Kitabe/blob/master/mainapp/helpers.py#L187)

**Describe alternatives you've considered**
If there is any other alternative you would like to suggest go ahead.

**Additional context**
Concerning tests should be written.

**Comment by saaalik** [comment]
I do like the idea of dropping down on TF-IDF recommendations and only taking top 3-4 with highest relevancy score but adding in books of "Similar" genre in recommendations would make things very complicated

A word does not have to associate with any kind of genre. For example, If you search for 'three' in the search box
The 5 results returned are:
three cups of tea - `Memior` `Biography`
the drawing of the three - `Novel` `Horror fiction` `Dark fantasy` `Fantasy Fiction` `Science Fiction` `Western fiction`
the true story of three little pigs - `Children's literature` `Picture book` `Fiction`
three to get deadly - `Mystery` `Crime novel` `Crime Fiction`
janet evanovich three and four book set - `Not even a book`
and figuring out a 'common' genre in the following would be too far fetched,...
Unless we meke the user specify the genre they are searching for... like using a check box and then searching

**Comment by Praful932** [comment]
Hi @saaalik 👋 Thanks for your interest in this project. I believe you are confusing between search and recommendations here. For search a simple string match is done, just like you do Search anywhere else.
Regarding your concern regarding TF-IDF, if you look at the notebooks, We not have only considered Book Title but also Author Name and Rating while recommending similar books using this approach. 
So in short TF-IDF and string match are completely different things. You can look at the Notebooks in the readme for clarity.
Currently we are not recommending on genre tag directly, so that is what I meant by adding 2 similar genre books when inshort of recommendations.

**Comment by saaalik** [comment]
ooh, well this is embarassing... So the helper.py is the only script being used in the recommendation part or are their any other python depenecies? I want the recommendations section's working to sink in as soon as possible. I am interested in working on this project.

**Comment by Praful932** [comment]
@saaalik Yep that's the only script and some similarity files are used. Check out the Project Structure and the Model exploration part specifically from [here](https://github.com/Praful932/Kitabe#contents) to understand how it's all put together. If you need any help, feel free to shoot doubts 🚀

**Comment by saaalik** [comment]
Updata: I spent the last two days scribbling a mind map connecting all the csv, pickle, npz etc files to their respective py files to better understand things at work. Only the files "Cleaning_and_Embedding_Notebook.pynb" and "Genre_wise_and_Count_vectorizer.pynb" do the count vectorizing and calculating cosine similarities on the data and dump data sets to be manipulated further. I still haven't been able to wrap my head around the working of these two notebooks completely, so it might take me a little longer to complete, but working with data on such large scale (large for me atleast) is really exhilerating, will complete it asap

**Comment by Praful932** [comment]
Glad to see your involvement, you're picking things up fast🚀, You can shoot doubts if you encounter any 🙂

**Comment by Praful932** [comment]
@saaalik Any updates?

**Comment by saaalik** [comment]
To print outputs of the jupyter files and work on the tf-idf script and similar-genre script, I proceeded to install jupyter notebooks on my system but I wasn't able to do the same for some reason, It took me 2 days to give up on jupyter. Now instead i copied the script in my system locally with all the drive files used and printed results on each step and tried changing some books.csv inputs to see how its effect. After that I haven't made any progress yet. I just have to make the tf-idf outputs and similar genre outputs get presented in equal ration instead of letting the "top books" get printed when there are not enough books to recommend right?

**Comment by Praful932** [comment]
@saaalik You can start a [discussion](https://github.com/Praful932/Kitabe/discussions/new) if you are facing any trouble or reach out on project discord :)
yes if there are 2n remaining books, **take n from top n tf-idf** and **take another top n from majority genre** from user's already rated books and the recommendations

**Comment by Praful932** [comment]
@saaalik Any updates?

**Comment by saaalik** [comment]
I am making changes in this snippet
```python
if len(best_bookids) < recommendations:
        # If not enough recommendations
        best_bookids = best_bookids + cv_bookids[3:(recommendations + 3 - len(best_bookids))]
```
here 2n = ( recommendations - len(best_bookids) )
then, I am converting it to:
```python
if len(best_bookids) < recommendations:
        # If not enough recommendations
        two_n = (recommendations - len(best_bookids))
        tf_idf_bookids = cv_bookids[3: (two_n-3)//2]
        sim_genre_bookids = most_common_genre_books(book_ids)   #BOOK IDS LIST containing books of the most common genre
        best_bookids =  best_bookids + tf_idf_bookids + sim_genre_bookids
```

**Comment by Praful932** [comment]
@saaalik Could you explain the approach, and add the comment above the code not beside

**Comment by saaalik** [comment]
Okey, so the variable `two_n` stores the total number of books remained to be recommended, out of which
half of the bookids would be retrieved from remaining cv_bookids - `tf_idf_bookids`
and other half from other `sim_genre_bookids` which will take two lists as parameters
1. `already_rated_bookids = {bookid1, bookid2, bookid3}` (The `already_rated_bookids` variable currently stores book ratings which doesn't make any sense, so I have fixed it, please correct me if it was supposed to stay that way)
2. `already_recommended_bookids = tf_idf_bookids + best_bookids = {bookid1, bookid4, bookid5}`
Add the two lists into 1 single set, loop through each bookid and find the most common genre among all books and return `n` remaining bookids.

**Comment by Praful932** [comment]
You're correct for fixing `1.`
```python
cv_bookids[3: (two_n+3)//2]
```
I think there should be a `+` here.
For `2.` Better would be finding common genre in already rated books as well instead of just recommendations
Otherwise all feels good to me

**Comment by subhangi2731** [comment]
@Praful932 please assign this issue to me

**Comment by Praful932** [comment]
Fixed by #135
